catris.py

```diff
@@ -928,55 +936,67 @@ def _add_high_score(self, file_name: str, hs: HighScore) -> list[HighScore]:
-                    players=[p.name for p in game.players],
-                )
-                print("Game over!", high_score)
-                game.players.clear()
+        game = self.__games[game_class]
+        assert game.is_valid()
+        assert not game.game_is_over()
+        yield game
+
+        assert game.is_valid()
+        if game.game_is_over():
+            duration_ns = time.monotonic_ns() - game.start_time
+            high_score = HighScore(
+                score=game.score,
+                duration_sec=duration_ns / (1000 * 1000 * 1000),
+                players=[p.name for p in game.players],
+            )
+            print("Game over!", high_score)
+            game.players.clear()
 
-                playing_clients = [
-                    c for c in self.clients if isinstance(c.view, PlayingView)
-                ]
+            playing_clients = [
+                c for c in self.clients if isinstance(c.view, PlayingView)
+            ]
 
-                assert render
-                if playing_clients:
-                    all_high_scores = self._add_high_score(
-                        game.HIGH_SCORES_FILE, high_score
-                    )
-                    all_high_scores.sort(key=(lambda hs: hs.score), reverse=True)
-                    best5 = all_high_scores[:5]
-                    for client in playing_clients:
-                        client.view = GameOverView(
-                            client, type(game), high_score, best5
-                        )
-                        client.render()
-                else:
-                    print("Not adding high score because everyone disconnected")
+            assert render
+            if playing_clients:
+                for client in playing_clients:
+                    client.view = GameOverView(client, type(game), high_score)
+                    client.render()
+                asyncio.create_task(self._high_score_task(type(game), high_score))
+            else:
+                print("Not adding high score because everyone disconnected")
 
-            elif render:
-                for client in self.clients:
-                    if isinstance(client.view, (PlayingView, ChooseGameView)):
-                        client.render()
+        elif render:
+            for client in self.clients:
+                if isinstance(client.view, (PlayingView, ChooseGameView)):
+                    client.render()
 
     # TODO: instantiate new Game instead of resetting, so won't need to pass start_time
-    def _countdown(self, player: Player, original_game: Game, start_time: int) -> None:
+    async def _countdown(
+        self, player: Player, original_game: Game, start_time: int
+    ) -> None:
         while True:
-            time.sleep(1)
+            await asyncio.sleep(1)
             with self.access_game(type(original_game)) as game:
                 if game.start_time != start_time:
                     return
```

Comment:
A few things could be improved in future PRs, this could use canceling tasks for example.